File: travel-planner-frontend/src/services/authService.js
import axios from 'axios';
import { User } from '../models/User';

const BASE_URL = import.meta.env.VITE_USER_SERVICE_URL;

export const authService = {
    async register(name, email, password) {
        const response = await axios.post(`${BASE_URL}/api/users/register`, {
            name, email, password
        });
        return response.data;
    },

    async login(email, password) {
        const response = await axios.post(`${BASE_URL}/api/users/login`, {
            email, password
        });
        return response.data;
    },

    async getUser(id) {
        const response = await axios.get(`${BASE_URL}/api/users/${id}`);
        const u = response.data;
        return new User(u.id, u.name, u.email, u.role, u.createdAt);
    },

    async getAllUsers(token) {
        const response = await axios.get(`${BASE_URL}/api/users`, {
            headers: { Authorization: `Bearer ${token}` }
        });
        return response.data.map(u => new User(
            u.id, u.name, u.email, u.role, u.createdAt
        ));
    },

    async updateUserRole(id, role) {
        const response = await axios.put(`${BASE_URL}/api/users/${id}/role`, role, {
            headers: { 'Content-Type': 'application/json' }
        });
        const u = response.data;
        return new User(u.id, u.name, u.email, u.role, u.createdAt);
    },

    async deleteUser(id, token) {
        const response = await axios.delete(`${BASE_URL}/api/users/${id}`, {
            headers: { Authorization: `Bearer ${token}` }
        });
        return response.data;
    }
};

File: UserService/Services/AuthService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.IdentityModel.Tokens;
using Microsoft.ServiceFabric.Services.Remoting.Client;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using TravelPlanner.Shared.Interfaces;
using UserService.Data;
using UserService.DTOs;
using UserService.Models;
using Microsoft.ServiceFabric.Services.Remoting.Client;
using TravelPlanner.Shared.Interfaces;

namespace UserService.Services
{
    public class AuthService : IAuthService
    {
        private readonly AppDbContext _context;
        private readonly IConfiguration _configuration;

        public AuthService(AppDbContext context, IConfiguration configuration)
        {
            _context = context;
            _configuration = configuration;
        }

        public async Task<string> Register(RegisterDto dto)
        {
            var existingUser = await _context.Users
                .FirstOrDefaultAsync(u => u.Email == dto.Email);

            if (existingUser != null)
                throw new Exception("User with this email already exists.");

            var user = new User
            {
                Name = dto.Name,
                Email = dto.Email,
                PasswordHash = BCrypt.Net.BCrypt.HashPassword(dto.Password),
                Role = "User"
            };

            _context.Users.Add(user);
            await _context.SaveChangesAsync();

            return GenerateToken(user);
        }

        public async Task<string> Login(LoginDto dto)
        {
            var user = await _context.Users
                .FirstOrDefaultAsync(u => u.Email == dto.Email);

            if (user == null || !BCrypt.Net.BCrypt.Verify(dto.Password, user.PasswordHash))
                throw new Exception("Invalid email or password.");

            return GenerateToken(user);
        }

        public async Task<UserDto> GetUser(int id)
        {
            var user = await _context.Users.FindAsync(id);
            if (user == null)
                throw new Exception("User not found.");
            return MapToDto(user);
        }

        private string GenerateToken(User user)
        {
            var key = new SymmetricSecurityKey(
                Encoding.UTF8.GetBytes(_configuration["Jwt:Key"]));

            var credentials = new SigningCredentials(
                key, SecurityAlgorithms.HmacSha256);

            var claims = new[]
            {
                new Claim(ClaimTypes.NameIdentifier, user.Id.ToString()),
                new Claim(ClaimTypes.Email, user.Email),
                new Claim(ClaimTypes.Role, user.Role)
            };

            var token = new JwtSecurityToken(
                issuer: _configuration["Jwt:Issuer"],
                audience: _configuration["Jwt:Audience"],
                claims: claims,
                expires: DateTime.UtcNow.AddDays(7),
                signingCredentials: credentials
            );

            return new JwtSecurityTokenHandler().WriteToken(token);
        }

        private static UserDto MapToDto(User user)
        {
            return new UserDto
            {
                Id = user.Id,
                Name = user.Name,
                Email = user.Email,
                Role = user.Role,
                CreatedAt = user.CreatedAt
            };
        }

        public async Task<List<UserDto>> GetAllUsers()
        {
            return await _context.Users
                .Select(u => MapToDto(u))
                .ToListAsync();
        }

        public async Task<UserDto> UpdateUserRole(int id, string role)
        {
            var user = await _context.Users.FindAsync(id);
            if (user == null)
                throw new Exception("User not found.");

            if (role != "User" && role != "Admin")
                throw new Exception("Role must be User or Admin.");

            user.Role = role;
            await _context.SaveChangesAsync();
            return MapToDto(user);
        }

        public async Task DeleteUser(int id)
        {
            var user = await _context.Users.FindAsync(id);
            if (user == null)
                throw new Exception("User not found.");

            var tripNotifier = ServiceProxy.Create<ITripNotifier>(
                new Uri("fabric:/TravelPlannerApp/TripService"));

            await tripNotifier.DeleteTripsByUser(id);

            _context.Users.Remove(user);
            await _context.SaveChangesAsync();
        }
    }
}